vms/platformvm/spend.go

```diff
@@ -123,46 +131,46 @@ func (vm *VM) stake(
 		// The remaining value is initially the full value of the input
 		remainingValue := in.Amount()
 
-		// Stake any value that should be staked
-		amountToStake := math.Min64(
-			amount-amountStaked, // Amount we still need to stake
-			remainingValue,      // Amount available to stake
+		// Spend any value that should be spent
+		amountToSpend := math.Min64(
+			totalAmountToSpend-amountSpent, // Amount we still need to spend
+			remainingValue,                 // Amount available to spend
 		)
-		amountStaked += amountToStake
-		remainingValue -= amountToStake
+		amountSpent += amountToSpend
+		remainingValue -= amountToSpend
 
 		// Add the input to the consumed inputs
 		ins = append(ins, &avax.TransferableInput{
 			UTXOID: utxo.UTXOID,
 			Asset:  avax.Asset{ID: vm.ctx.AVAXAssetID},
-			In: &StakeableLockIn{
-				Locktime:       out.Locktime,
+			In: &LockedIn{
+				LockIDs:        lockedOut.LockIDs,
 				TransferableIn: in,
 			},
 		})
 
-		// Add the output to the staked outputs
-		stakedOuts = append(stakedOuts, &avax.TransferableOutput{
+		// Add the output to the transitioned outputs
+		outs = append(outs, &avax.TransferableOutput{
```

Comment:
Improvement: Instead 1 out for each in, create a map[key]{amount, outputowners, !LockstateTxId} where key is the hash of outputowners / !LockstateTx. After this loop process outs by using map for compacted outs.